core/modules/replica/JobController.cc

```diff
@@ -0,0 +1,503 @@
+////////////////////////////////////////////////////////////////////
+
+JobController::Ptr JobController::create(ServiceProvider::Ptr const& serviceProvider) {
+    return JobController::Ptr(
+        new JobController(serviceProvider));
+}
+
+JobController::JobController(ServiceProvider::Ptr const& serviceProvider)
+    :   _serviceProvider(serviceProvider),
+        _controller(Controller::create (serviceProvider)),
+        _stop(false) {
+}
+
+void JobController::run() {
+
+    LOGS(_log, LOG_LVL_DEBUG, "JobController::run");
+
+    util::Lock lock(_mtx, "JobController::run");
+
+    if (not isRunning()) {
+
+        // Run the controller in its own thread.
+
+        _controller->run();
+
+        JobController::Ptr self = shared_from_this();
+
+        _thread.reset(
+            new std::thread(
+                [self] () {
+
+                    // This will prevent the scheduler from existing unless
+                    // instructed to do so
+
+                    util::BlockPost blockPost(0, 1000); // values of parameters are meaningless
+                                                        // in this context because the object will
+                                                        // be always used to wait for a specific interval
+
+                    unsigned int const wakeUpIvalMillisec =
+                        1000 * self->_serviceProvider->config()->jobSchedulerIvalSec();
+
+                    while (blockPost.wait(wakeUpIvalMillisec)) {
+
+                        // Initiate the stopping sequence if requested
+                        if (self->_stop) {
+
+                            // Cancel all outstanding jobs
+                            self->cancelAll();
+
+                            // Block here waiting before the controller will stop
+                            self->_controller->stop();
+
+                            // Quit the thread
+                            return;
+                        }
+
+                        // Check if there are jobs scheduled to run on the periodic basis
+                        self->runScheduled();
+                    }
+                }
+            )
+        );
+    }
+}
+
+bool JobController::isRunning() const {
+    return _thread.get() != nullptr;
+}
+
+void JobController::stop() {
+
+    LOGS(_log, LOG_LVL_DEBUG, "JobController::stop");
+
+    // IMPORTANT:
+    //
+    //   Never attempt running these operations within util::Lock lock(_mtx)
+    //   due to a possibile deadlock when asynchronous handlers will be
+    //   calling the thread-safe methods. A problem is that until they finish
+    //   in a clean way the thread will never finish, and the application will
+    //   hang on _thread->join().
+
+    // util::Lock lock(_mtx)  (disabled)
+
+    if (not isRunning()) return;
+
+    // Tell the thread to finish
+    _stop = true;
+
+    // Join with the thread before clearning up the pointer
+    _thread->join();
+    _thread.reset(nullptr);
```

Comment:
if you don't want to hold the lock because of possible deadlocks, then more than 1 caller can be in here at the same time? Couldn't one caller reset `_thread` here before the other caller gets to`_thread->join()`? That would be bad...

Reply:
The JobController has been redesigned. It's much simpler now. And it won't start its own thread.